### ISSUE TITLE
Questions on verifying "precomputed" dataset

### ISSUE THREAD

**Issue by manoaman** [google/neuroglancer#204]
Hi,

I'm trying to load a series of tiffs, precomputed with CloudVolume, and trying to load into Neuroglancer but can't seem to figure out why the viewer is not presenting the data. Either given resolution parameters are not accurate in CloudVolume settings (info), or issues with bright/contrast control, or something else. (Screenshot attached.)

I also realized there is a way to control bright/contrast from JSON entry.  Do you know a good example of a json file to control bright/contrast in Neuroglancer?  I can seem to figure out how to control this.

> The fragment shader code can be entered interactively from the side panel for an image layer, or programmatically by specifying >a 'shader' property of the JSON specification of the image layer.https://github.com/google/neuroglancer/blob/master/src/neuroglancer/sliceview/image_layer_rendering.md

Any advise would be appreciated.

Thank you,
Seita

![Screen Shot 2020-03-06 at 11 41 40 AM-1](https://user-images.githubusercontent.com/47464840/76139382-08ae5580-6004-11ea-8898-db01666cd36e.png)

**Comment by jbms** [comment]
As far as the data not loading, you might want to check the console or the Network panel of your browser developer tools to see if there are any failed requests.  If you are able to provide me with access to the volume, I can also take a look myself.

As far as controlling brightness and contrast, there aren't built-in options for that but you can specify a shader to do that, in the layer side panel.  You can then set the uicontrol values from the JSON state:

```
#uicontrol vec3 color color(default="red")
#uicontrol float brightness slider(min=-1, max=1)
#uicontrol float contrast slider(min=-3, max=3, step=0.01)
void main() {
  emitRGB(color *
          (toNormalized(getDataValue()) + brightness) *
          exp(contrast));
}
```

**Comment by manoaman** [comment]
Hi jbms,

I had doubts with "info" parameters with CloudVolume especially with resolution and that seemed to have done the trick.  However, it is still coming up black so I'm doubting contrast/brightness need to be adjusted because viewing original tiffs in Fiji(ImageMagick) required to change contrast/brightness to view.  (Not sure if there is still an issue with CloudVolume info values need to be revisited.  Is there an easier way to generate this info?)  

I also discovered in a different post that SWC files can be loaded directly to Neuroglancer.  Is there a document or steps I can read how to accomplish this?

Thanks again for your help.

**Comment by manoaman** [comment]
Tried adjusting brightness/contrast/opacity on the Neuroglancer but that didn't seem to help so I'm stuck where it's failing to load the data.

I don't have enough storage to put the entire tiff stack but here is shared drive if you can take a look at.  It contains precomputed dataset generated from CloudVolume.

https://drive.google.com/drive/folders/1X2Vebe9uBFXTY-IjqQQUPmyIaIhUPZS0?usp=sharing

I would appreciate if you could share insights on where to investigate.  Thanks so much.


![Screen Shot 2020-03-10 at 5 21 27 PM](https://user-images.githubusercontent.com/47464840/76370308-abf8b680-62f3-11ea-9571-22a25eab0f0b.png)
`{
  "dimensions": {
    "x": [
      0.01309425,
      "m"
    ],
    "y": [
      0.02506075,
      "m"
    ],
    "z": [
      0.000219,
      "m"
    ]
  },
  "position": [
    2013.8824462890625,
    3846.810791015625,
    193.0858612060547
  ],
  "crossSectionScale": 0.0014269406392694065,
  "projectionOrientation": [
    -0.18260467052459717,
    -0.2293349951505661,
    0.6760212182998657,
    0.6760594248771667
  ],
  "projectionScale": 93.51598173515983,
  "layers": [
    {
      "type": "image",
      "source": {
        "url": "precomputed://http://localhost:8888/precomputed/cloud_volume_test/precomputed",
        "subsources": {
          "default": true,
          "bounds": true
        },
        "enableDefaultSubsources": false
      },
      "opacity": 1,
      "blend": "default",
      "shader": "#uicontrol vec3 color color(default=\"white\")\n#uicontrol float min slider(default=0, min=0, max=1, step=0.01)\n#uicontrol float max slider(default=1, min=0, max=1, step=0.01)\n#uicontrol float brightness slider(default=0, min=-1, max=1, step=0.1)\n#uicontrol float contrast slider(default=0, min=-3, max=3, step=0.1)\n\nfloat scale(float x) {\n  return (x - min) / (max - min);\n}\n\nvoid main() {\n  emitRGB(\n    color * vec3(\n      scale(\n        toNormalized(getDataValue()))\n       + brightness) * exp(contrast)\n  );\n}",
      "shaderControls": {
        "color": "#fff7f6",
        "min": 0.67,
        "max": 0.7,
        "brightness": 0.4,
        "contrast": 1.7
      },
      "name": "test"
    }
  ],
  "selectedLayer": {
    "layer": "test",
    "visible": true,
    "size": 552.5703125
  },
  "layout": "xy",
  "statistics": {
    "visible": true,
    "size": 121.6484375
  }
}`

**Comment by jbms** [comment]
From your screenshot it appears that the data is loading correctly since a value for the position under the mouse is displayed in the layer bar.

With the shader you are using, min and max are specified as fractions of the full range of [0, 65535], so you may just need to set much smaller values of min and max, e.g. min = 0 and max = 0.003

**Comment by manoaman** [comment]
That did the trick.  But I can't seem to navigate through the stacks.  I'm still new to cross sectional view.  Do you mind advising how to navigate z-stacks in Neuroglancer?  Also, what are min/max values?  Thanks!

![Screen Shot 2020-03-13 at 10 24 08 AM](https://user-images.githubusercontent.com/47464840/76644885-1b99bc00-6515-11ea-92c8-fc22317dc180.png)

**Comment by jbms** [comment]
Click the ? icon in the top right to see the key/mouse bindings.  You can also click on the  x/y/z position at the top left to open a slider that lets you move in each dimension individually.

In general you can mouse drag like in google maps to translate within a cross section, and use the mouse wheel to move orthogonal to a cross section.

Also, your screenshot shows an x resolution of 1 pixel = 1.3 cm, y resolution of 1 pixel = 2.5cm, z resolution of 1 pixel 219um.  That is why in the xz and yz cross sections the full extent of the z dimension is just 1 pixel.  Are you sure that is correct?

**Comment by manoaman** [comment]
Great that help popup is nice.  No, the volume metrics has to be corrected and I think 810um x 1550um x 219 um is more accurate value I should be configuring.  Could be corrected by modifying the following "info" file?  Or do I also need to run CloudVolume to regenerate the dataset again?

Also, I realized some tiles are not loading up as you see in the screenshot.  How do I pick the number of "chunk_sizes" or is there another parameters I should be looking at?

```
% /usr/local/ImageMagick-6.6.8_64bit/bin/identify -format "%w x %h %x x %y" C0_Z050.tif
4029 x 7711 49759.2 PixelsPerCentimeter x 49758.1 PixelsPerCentimeter

```
`{
  "data_type": "uint16",
  "num_channels": 1,
  "scales": [
    {
      "chunk_sizes": [
        [
          1024,
          1024,
          1
        ]
      ],
      "encoding": "raw",
      "key": "810000_1550000_219000",
      "resolution": [
        810000,
        1550000,
        219000
      ],
      "size": [
        810,
        1550,
        219
      ],
      "voxel_offset": [
        0,
        0,
        1
      ]
    }
  ],
  "type": "image"`

![Screen Shot 2020-03-13 at 5 21 34 PM](https://user-images.githubusercontent.com/47464840/76670687-21ac8e80-654f-11ea-935d-5f5737bbbf80.png)

**Comment by jbms** [comment]
Yes, you can correct the pixel size just by modifying the info file.

As for why some tiles may not be loading, there might be additional information in the browser developer console.

**Comment by manoaman** [comment]
hi jbms, so I think I managed to fix pixel size from info file but there are still problems I'm seeing with data conversion and loading files.  And I have these questions.

1. Scale does not seem to reflect the actual size.  Where do I control this in JSON file? e.g.) 1mm in the screenshot.

![Screen Shot 2020-03-18 at 3 51 37 PM](https://user-images.githubusercontent.com/47464840/77014857-76298280-6930-11ea-8bfa-84b1741478b9.png)

2. Looking at chrome dev tool, some tiles are not found and CloudVolume script didn't seem to generate them.  Is there a place in CloudVolume or info parameters I can revisit?

![Screen Shot 2020-03-18 at 3 48 03 PM](https://user-images.githubusercontent.com/47464840/77014863-775aaf80-6930-11ea-92bb-11a5d96c3dee.png)

3. What should be the chunk size in info?  Currently it's set to chunk_size=[1024, 1024, 1].  Should it be smaller chunks?

Thanks for you help!

I'm also attaching json config from Neuroglancer.

`{
  "dimensions": {
    "x": [
      0.00081,
      "m"
    ],
    "y": [
      0.00155,
      "m"
    ],
    "z": [
      0.000219,
      "m"
    ]
  },
  "position": [
    1752,
    3970,
    183.27389526367188
  ],
  "crossSectionScale": 0.04566210045662101,
  "projectionOrientation": [
    0.03876139223575592,
    -0.8888232111930847,
    0.18656191229820251,
    0.41675591468811035
  ],
  "projectionScale": 766082.9223744293,
  "layers": [
    {
      "type": "image",
      "source": {
        "url": "precomputed://http://localhost:8888/precomputed/path_to_files",
        "subsources": {
          "default": true,
          "bounds": true
        },
        "enableDefaultSubsources": false
      },
      "opacity": 1,
      "blend": "default",
      "shader": "#uicontrol vec3 color color(default=\"white\")\n#uicontrol float min slider(default=0, min=0, max=1, step=0.01)\n#uicontrol float max slider(default=1, min=0, max=1, step=0.01)\n#uicontrol float brightness slider(default=0, min=-1, max=1, step=0.1)\n#uicontrol float contrast slider(default=0, min=-3, max=3, step=0.1)\n\nfloat scale(float x) {\n  return (x - min) / (max - min);\n}\n\nvoid main() {\n  emitRGB(\n    color * vec3(\n      scale(\n        toNormalized(getDataValue()))\n       + brightness) * exp(contrast)\n  );\n}\n",
      "shaderControls": {
        "color": "#531352",
        "max": 0.003,
        "brightness": 0.6,
        "contrast": -0.6
      },
      "crossSectionRenderScale": 131.82416447567192,
      "name": "test"
    }
  ],
  "selectedLayer": {
    "layer": "test",
    "visible": true,
    "size": 426
  },
  "layout": "4panel",
  "statistics": {
    "visible": true,
    "size": 153
  }
}`

**Comment by jbms** [comment]
Can you explain what you mean by scale does not reflect the actual size?

The chunk_size in the info file must match how you exported the data into precomputed format.  If you only intend to view xy cross sections it may make sense to use a 2-d chunk size like 512, 512, 1 (or 1024, 1024, 1 as you have now).  Otherwise you may want to use a 3-d chunk size, e.g. 64, 64, 64

**Comment by manoaman** [comment]
I added a red arrow in the screenshot.  It's currently showing 5m but the actual width of this image is 810um.  So I'm expecting something like 1mm.  How do I configure this correctly?

![Screen Shot 2020-03-19 at 9 12 44 AM](https://user-images.githubusercontent.com/47464840/77088972-707c7d00-69c2-11ea-90e1-3102dc98a3a6.png)

**Comment by jbms** [comment]
In the screenshot, the x resolution is listed as 1 pixel = 810um.  The resolution field in the info file should specify the size in nanometers of a single pixel, not of the entire volume.

**Comment by manoaman** [comment]
I see, resolution needs to set per pixel in nanometers and not the entire volume.  For 3d chunk size with 64,64,64, I seem to be getting a conversion error.  Do tiff stacks work in the first place?

```
Could not convert data to an integer.

  Non-Aligned writes are disabled by default. There are several good reasons 
  not to use them. 

  1) Memory and Network Inefficiency
    Producing non-aligned writes requires downloading the chunks that overlap
    with the write area but are not wholly contained. They are then painted
    in the overlap region and added to the upload queue. This requires
    the outer shell of the aligned image to be downloaded painted, and uploaded. 
  2) Race Conditions
    If you are using multiple instances of CloudVolume, the partially overlapped 
    chunks will be downloaded, partially painted, and uploaded. If this procedure
    occurs close in time between two processes, the final chunk may be partially
    painted. If you are sure only one CloudVolume instance will be accessing an
    area at any one time, this is not a problem.

  If after reading this you are still sure you want non-aligned writes, you can
  set non_aligned_writes=True.

  Alignment Check: 
    Mip:             0
    Chunk Size:      [64 64 64]
    Volume Offset:   [0 0 1]
    Received:        Bbox([0, 0, 6],[4029, 7711, 7], dtype=int32) 
    Nearest Aligned: Bbox([0, 0, 1],[4029, 7711, 65], dtype=int32)

Processing C0_Z009.tif
Could not convert data to an integer.
```

**Comment by jbms** [comment]
Questions about cloudvolume should be directed at the authors of that package.

It may be easiest to stick with a 1024x1024x1 or 512x512x1 chunk size, though.

**Comment by manoaman** [comment]
Ok, will redirect question to cloudvolume too.  (I can't seem to get response right away.)  In the meantime, could you verify if I am specifying the resolution correctly? thanks

> 4029 x 7711 49759.2 PixelsPerCentimeter x 49758.1 PixelsPerCentimeter

>   Format: TIFF (Tagged Image File Format)
>   Class: DirectClass
>   Geometry: 4029x7711+0+0
>   Resolution: 49759.2x49758.1
>   Print size: 0.0809699x0.15497
>   Units: PixelsPerCentimeter

And each z-stack is 1um with 219 series.

This is the result of output of what I'm getting from ImageMagick.  If I'm supposed to specify "the size in nanometers of a single pixel",

X: 1px = 0.00497522 nm
Y: 1px = 0.00497581 nm
Z: 1px = 0.02485 nm

```
      "resolution": [
        0.00497522,
        0.00497581,
        0.02485
      ],
      "size": [
        4029,
        7711,
        219
      ],
```

**Comment by jbms** [comment]
The resolution in the info file is specified in nanometers per pixel, not pixels per nanometer, so you should invert the values for x and y, i.e. an x and y resolution of about 200 nm/pixel.  If each z section is 1um, you should specify a resolution value of 1000 nm/pixel for z.

**Comment by manoaman** [comment]
So it looks promising with proper configuration in info. Thanks so much for the clarification.  Although, I'm still battling with some tiles not fully loaded when navigating through z-stack.  Originally, I had 72MB of a tiff, and downsized to 18MB.  That did improve the loading so I still feel latency and size of the files may be causing this.  I may try further downsize the original files.  What is the suggested size or resolution for loading in general?

**Comment by manoaman** [comment]
Hi jbms,

What is the advantage over using 3-d chunk size over 2-d?  It seems to me that I can view xz and yz cross section just fine on 2-d chunk size.  Resolution seems a bet rough but I couldn't tell the difference.  Could it be related to efficient data transferring over the network?

> If you only intend to view xy cross sections it may make sense to use a 2-d chunk size like 512, 512, 1 (or 1024, 1024, 1 as you have now). Otherwise you may want to use a 3-d chunk size, e.g. 64, 64, 64

**Comment by jbms** [comment]
Yes, the chunk size affects the efficiency. 

For example, if your chunk size is (1024, 1024, 1), then if you view an xz or yz cross-section, you end up having to load 1024x as much data as is actually shown at one time, e.g. if you had an xz (1024, 1, 1024) or yz (1, 1024, 1024) chunk.

In contrast, if you use a (64, 64, 64) chunk, then you only have to load 64x as much data as is actually shown at one time.

**Comment by manoaman** [comment]
I used igneous to rechunk single slice chunks with (64, 64, 64) chunk; loaded on Neuroglancer does not seem to show the results I saw from single slice chunks.  And I hope you can answer to these questions.  First, the chunks seem not available for the latter part of x-axis. Could this be a failure of rechunking single slice chunks?  Or json configuration needs to be corrected?  Second, why is 1600-1600-1000 downsampled level always called and not other downsampled levels?  (Could it be just editing info file and choose the "scales" level?)  Thanks!

![Screen Shot 2020-04-06 at 8 36 53 PM](https://user-images.githubusercontent.com/47464840/78627499-6d98dd80-7846-11ea-851a-b052c2cfe348.png)

![Screen Shot 2020-04-06 at 8 28 05 PM](https://user-images.githubusercontent.com/47464840/78627076-72a95d00-7845-11ea-8988-f1fb8b4f4c94.png)

[info.txt](https://github.com/google/neuroglancer/files/4441886/info.txt)

[json.txt](https://github.com/google/neuroglancer/files/4441885/json.txt)

**Comment by jbms** [comment]
The reason you are only seeing the 1600_1600_1000 source is that you have the "crossSectionRenderScale" set much higher than the default of 1px.  Go to the "Rendering" tab in the layer side panel and double click on the "Resolution (slice)" control to reset it to 1.  Then if you zoom in you should see the other sources loaded.

The chunk statistics does show 11 failed chunk requests.  If they are indeed missing, that would be something to ask the igneous maintainers about.

**Comment by manoaman** [comment]
Yes, once resetting "Resolution (slice)", I do see other sources active in visible chunks monitor on the bottom panel.  It seems when I zoom-in, rough resolution could be made a bet smoother.  What is the general use of this UI control?

I still feel zoomed-in images seem rough in resolution.  Are there other parameters I need to tweak on Neuroglancer to make the visual images clearer? Or could this be a topic on rechunking with Igneous?

![Screen Shot 2020-04-07 at 11 37 49 AM](https://user-images.githubusercontent.com/47464840/78706828-8ea01380-78c4-11ea-800f-7b55738eec67.png)


![Screen Shot 2020-04-07 at 11 38 36 AM](https://user-images.githubusercontent.com/47464840/78706809-88119c00-78c4-11ea-80e9-fe4da888df7c.png)
![Screen Shot 2020-04-07 at 11 38 17 AM](https://user-images.githubusercontent.com/47464840/78706823-8d6ee680-78c4-11ea-9b28-ca7091bffb4b.png)

**Comment by jbms** [comment]
The "Resolution" UI controls allow you to limit the resolution that is used --- e.g. if the data is slow to access you could limit it to a coarser resolution.  However, in practice I'd say it is primarily useful as a display element to show the resolutions that are used and the progress of downloading chunks, rather than as a control.

As far as it appearing "rough" when zoomed in, if you zoom in extremely far, you can compare the scale bar to the size of a single voxel to confirm that you are seeing the finest resolution data.

Neuroglancer uses nearest neighbor rather than linear interpolation by default, which means you see individual voxels as blocks when you zoom in far.  You can enable linear interpolation by changing "getDataValue" to "getInterpolatedDataValue" in the shader code.  This only operates within chunks, not across chunks, though, so if you have 2-d chunks then you will not get any interpolation in z.  With your rechunked volume you will get interpolation in z, though.

**Comment by manoaman** [comment]
Did I miss anything with replacing "getDataValue" to "getInterpolatedDataValue" ?

![Screen Shot 2020-04-07 at 12 59 31 PM](https://user-images.githubusercontent.com/47464840/78714093-20614e00-78d0-11ea-934a-9801b5b61444.png)
![Screen Shot 2020-04-07 at 12 59 24 PM](https://user-images.githubusercontent.com/47464840/78714098-222b1180-78d0-11ea-8053-d9550f762c48.png)

**Comment by jbms** [comment]
Which Neuroglancer instance/version are you using?  Perhaps it is a version from several months ago that doesn't support getInterpolatedDataValue yet?


See example below with current version:

https://neuroglancer-demo.appspot.com/#!%7B%22dimensions%22:%7B%22x%22:%5B8e-9%2C%22m%22%5D%2C%22y%22:%5B8e-9%2C%22m%22%5D%2C%22z%22:%5B8e-9%2C%22m%22%5D%7D%2C%22position%22:%5B3308.83642578125%2C3286.4208984375%2C4045%5D%2C%22crossSectionScale%22:0.09208897957928124%2C%22projectionOrientation%22:%5B0.31435418128967285%2C0.8142172694206238%2C0.4843378961086273%2C-0.06040274351835251%5D%2C%22projectionScale%22:4593.980956070107%2C%22layers%22:%5B%7B%22type%22:%22image%22%2C%22source%22:%22precomputed://gs://neuroglancer-public-data/flyem_fib-25/image%22%2C%22shader%22:%22void%20main%28%29%20%7B%5Cn%20%20emitGrayscale%28toNormalized%28getInterpolatedDataValue%28%29%29%29%3B%5Cn%7D%5Cn%22%2C%22name%22:%22image%22%7D%2C%7B%22type%22:%22segmentation%22%2C%22source%22:%22precomputed://gs://neuroglancer-public-data/flyem_fib-25/ground_truth%22%2C%22segments%22:%5B%22158571%22%2C%2221894%22%2C%2222060%22%2C%2224436%22%2C%222515%22%5D%2C%22name%22:%22ground-truth%22%2C%22visible%22:false%7D%5D%2C%22showSlices%22:false%2C%22selectedLayer%22:%7B%22layer%22:%22image%22%2C%22visible%22:true%2C%22size%22:554%7D%2C%22layout%22:%224panel%22%7D

**Comment by manoaman** [comment]
Good question.  How do I tell the version?  I think the code is from earlier this year (January).

**Comment by jbms** [comment]
The git commit or other version information unfortunately isn't stored in the built Neuroglancer, so you just have to keep track of which version was used to build it.  But in any case, it sounds like you need a more recent version if you want interpolation support.

**Comment by manoaman** [comment]
I'm still getting the same error after checking out the latest code from GitHub.  Is there a way to verify by looking at any of the .ts files?

**Comment by jbms** [comment]
Yeah, you can use git grep to search for getInterpolatedDataValue.

```shell
$ git grep getInterpolatedDataValue
src/neuroglancer/sliceview/volume/frontend.ts:${getShaderType(dataType)} getInterpolatedDataValue(${dataAccessChannelParams}) {
src/neuroglancer/sliceview/volume/frontend.ts:${getShaderType(dataType)} getInterpolatedDataValue() { return getInterpolatedDataValue(0); }
```

**Comment by manoaman** [comment]
Yes, I think I managed to get the latest code working.  Thank you.  

I see two new features in GUI which I didn't realize before.  What are the use cases of these features?  Does volume rendering work with precomputed chunks?

![Screen Shot 2020-04-08 at 10 58 50 AM](https://user-images.githubusercontent.com/47464840/78830384-b5ca1400-799c-11ea-838a-0f62ad6c2700.png)

![Screen Shot 2020-04-08 at 1 25 37 PM](https://user-images.githubusercontent.com/47464840/78830381-b498e700-799c-11ea-9dae-ce260952ca95.png)

**Comment by jbms** [comment]
The first screenshot shows the zoom and depth control.  It shows the current zoom level for each dimension and shows you to change the relative scaling between dimensions.  That is most useful when the dimensions do not all have the same units, e.g.  if there are both space and time dimensions.  

The depth limit can be adjusted with alt+wheel and controls the near/far clipping planes used.

For the cross section view, the depth limit applies to skeletons and annotations only.

For the 3d projection view, the depth limit applies to everything, and is particularly important for volume rendering.

Volume rendering works with all data sources, including precomputed chunks.  However, it is still experimental and has some rough edges, and generally requires a lot of tuning of the shader to get a reasonable result.

**Comment by jbms** [comment]
Here is an example of volume rendering on the FIB25 dataset.  You could copy the shader code and adapt to your own dataset:

https://neuroglancer-demo.appspot.com/#!%7B%22dimensions%22:%7B%22x%22:%5B8e-9%2C%22m%22%5D%2C%22y%22:%5B8e-9%2C%22m%22%5D%2C%22z%22:%5B8e-9%2C%22m%22%5D%7D%2C%22position%22:%5B4086.528076171875%2C3125.428466796875%2C4095.438232421875%5D%2C%22crossSectionScale%22:0.8672644965275753%2C%22projectionOrientation%22:%5B-0.6544808745384216%2C0.6523418426513672%2C-0.36523208022117615%2C-0.11274070292711258%5D%2C%22projectionScale%22:444.03942222211833%2C%22projectionDepth%22:-0.4240190079976629%2C%22layers%22:%5B%7B%22type%22:%22image%22%2C%22source%22:%22precomputed://gs://neuroglancer-public-data/flyem_fib-25/image%22%2C%22tab%22:%22source%22%2C%22name%22:%22image%22%7D%2C%7B%22type%22:%22image%22%2C%22source%22:%22precomputed://gs://neuroglancer-public-data/flyem_fib-25/image%22%2C%22opacity%22:1%2C%22shader%22:%22#uicontrol%20float%20brightness%20slider%28min=-10%2C%20max=10%2C%20default=0%2C%20step=0.1%29%5Cn#uicontrol%20float%20minValue%20slider%28min=0%2C%20max=255%2C%20default=0%2C%20step=1%29%5Cn#uicontrol%20float%20maxValue%20slider%28min=0%2C%20max=255%2C%20default=255%2C%20step=1%29%5Cn#uicontrol%20vec3%20color%20color%28default=%5C%22green%5C%22%29%5Cnvoid%20main%28%29%20%7B%5Cn%20%20float%20x%20=%20float%28getInterpolatedDataValue%28%29.value%29%3B%5Cn%20%20float%20val%20=%20%28x%20-%20minValue%29%20/%20%28maxValue%20-%20minValue%29%3B%5Cn%20%20float%20alpha%3B%5Cn%20%20if%20%28val%20%3C=%200.0%20%7C%7C%20val%20%3E=%201.0%29%20alpha%20=%200.0%3B%5Cn%20%20else%20alpha%20=%20val%3B%5Cn%20%20emitRGBA%28vec4%28color%2C%20alpha%2Aexp%28brightness%29%29%29%3B%5Cn%7D%5Cn%22%2C%22shaderControls%22:%7B%22minValue%22:90%2C%22maxValue%22:8%2C%22brightness%22:3.6%7D%2C%22volumeRendering%22:true%2C%22name%22:%22mito%22%7D%5D%2C%22showSlices%22:false%2C%22selectedLayer%22:%7B%22layer%22:%22mito1%22%2C%22visible%22:true%2C%22size%22:502%7D%2C%22layout%22:%7B%22type%22:%223d%22%2C%22orthographicProjection%22:true%7D%7D

**Comment by manoaman** [comment]
Zoom and depth control seem really nice.  It seems I can get much clear image view especially on the x-z, y-z planes.  I suppose ± values is F ratio?  I didn't quite catch the subtle differences in toggling check boxes.  What does toggling mean?

On the side note, sorry, this is much more general question on Neuroglancer.  I'm experiencing incomplete (?) chunk downloads and not sure why.  When I observe progressions of chunk downloads on Neuroglancer's download status panel, and Chrome Dev Tool, they seem to stop at certain point and I don't know what triggers.  (Togging to a different browser tab? Network speed? Laptop RAM usage?)  Is it looking at "M" and "G" columns for what chunks have been downloaded?  Is there a way to resume the downloads except another zoom-in/out?  What do you recommend on RAM size/network speed for viewing on a browser?

![Screen Shot 2020-04-10 at 11 09 40 AM](https://user-images.githubusercontent.com/47464840/79015214-374aaf00-7b21-11ea-99fb-ea90546ccce2.png)
![Screen Shot 2020-04-10 at 11 09 57 AM](https://user-images.githubusercontent.com/47464840/79015235-3c0f6300-7b21-11ea-8502-dae4f7f16aa3.png)
![Screen Shot 2020-04-10 at 11 18 46 AM](https://user-images.githubusercontent.com/47464840/79015241-3dd92680-7b21-11ea-9e1d-ddb6fd192050.png)

**Comment by manoaman** [comment]
Volume rendering looks really nice on your example.  I tried with my test images with same given shader controls but I couldn't tell if I am getting the best out of volume rendering.  Maybe needs more tweaking with deep control.

![Screen Shot 2020-04-10 at 11 32 09 AM](https://user-images.githubusercontent.com/47464840/79015333-65c88a00-7b21-11ea-9792-12f5d67a4d50.png)

**Comment by jbms** [comment]
The depth limit should not affect the display of volumetric data in the cross-section views --- are you seeing something different?  The +/- distance listed is the distance from the center position (the position listed at the top left of the neuroglancer window) to the near and far clipping planes, except that in perspective projection mode the near clipping plane may be closer to the center position if the listed distance would be impossible.  The far clipping plane is always exactly at the distance listed.

If you right click on the top bar (e.g. next to whether the position is listed), you can see the memory limits.  You can increase them, e.g. the GPU memory limit up to 2GB, but Chrome does not allow the process to use more than 4GB of memory in total.

For viewing multiple orthogonal planes, or for volume rendering, using 3d near-isotropic chunks will be much more efficient.

**Comment by manoaman** [comment]
I see. It could be chunks haven't fully loaded on the viewer. I thought it looked clearer when I alt+wheel on cross-section views.

> The depth limit should not affect the display of volumetric data in the cross-section views --- are you seeing something different?

This is very cool.  I didn't realize there is a feature to increase limit but seems like 2GB is as far as I can go.  Could this be the reason why I'm seeing unfilled chunks in cross-section views?

> If you right click on the top bar (e.g. next to whether the position is listed), you can see the memory limits. You can increase them, e.g. the GPU memory limit up to 2GB, but Chrome does not allow the process to use more than 4GB of memory in total.

I'm not familiar with 3d near-isotropic chunks.  Is this a different chunking method provided by CloudVolume?

**Comment by jbms** [comment]
By 3d isotropic chunks I just mean a chunk size where all dimensions are approximately the same length in physical units, e.g. if your voxel size is 200nm x 200nm x 1000nm , an example of an isotropic chunk size would be 100 x 100 x 20 voxels, because then the chunk is 20um x 20um x 20um.

**Comment by manoaman** [comment]
Thank you for the explanation and shed lights on where to tackle.  It seems like I can tweak CloudVolume to get better chunks for the visualization.

In the meantime, I realized I only have 1.5GB VRAM. (Not sure why Neuroglancer shows 2000000000 system memory limit.)  And the task manager shows it's reaching the limit of what I'm capable to run.  I think I should test with VRAM with 4GB to see if I see better results.

![Screen Shot 2020-04-13 at 11 21 44 AM](https://user-images.githubusercontent.com/47464840/79148334-2b0d5e80-7d7a-11ea-8ed3-b835f811af4e.png)

**Comment by jbms** [comment]
Neuroglancer defaults to 1GiB GPU memory and 2GiB system memory limits; however, these are just arbitrarily limits.  Neuroglancer has no way to determine how much memory your computer actually has.  You can adjust them, but the system memory limit should be at least as high as the GPU memory limit or there won't be any effect.  If you increase the limits too high, Chrome may kill the browser tab with an "Aw, snap" error, or you may get a Context lost error in the developer console (and the views will turn black), both of which will require reloading the page.

**Comment by manoaman** [comment]
I tried with a machine equipped with 4GiB GPU memory.  Going back to the question you had, it seems that I see these gray shapes missing tiles with depth zoom control. (Image attached.)

They appear in all cross section planes.  Does it mean I am too close to the image?  Do I control deep zoom by having a focus on 3D projection view?  (I can't tell if deep zoom is working on all the viewports or only on the focused viewport.)

> The depth limit should not affect the display of volumetric data in the cross-section views --- are you seeing something different?


![Screen Shot 2020-04-13 at 3 57 01 PM](https://user-images.githubusercontent.com/47464840/79179531-af80d100-7dbc-11ea-936f-795f39bd01cf.png)

![Screen Shot 2020-04-13 at 3 56 17 PM](https://user-images.githubusercontent.com/47464840/79179576-c3c4ce00-7dbc-11ea-857d-70e57a3cd7de.png)

**Comment by jbms** [comment]
Unfortunately it is hard for me to help fix this problem without a way to reproduce it myself.  If you are able to upload the data to a public server, or a public s3/cloud storage bucket, and send me a neuroglancer link, I would probably be able to help more easily.

To use google cloud storage, you can use the following commands to make the bucket usable from neuroglancer:

```shell
bucket=your-bucket-name
project=your-google-cloud-project-name

gsutil mb -p "${project}" "gs://$bucket"
gsutil iam ch allUsers:objectViewer "gs://$bucket"
gsutil cors set neuroglancer-cors-policy.json "gs://$bucket"
gsutil defacl set public-read "gs://$bucket"
```

where neuroglancer-cors-policy.json has the following contents:

```json
[{"maxAgeSeconds": 3600, "method": ["GET"], "origin": ["*"], "responseHeader": ["Content-Type", "Range"]}]
```

**Comment by manoaman** [comment]
That’s a good question. I could probably upload 21gb of precomputed data to google storage. But I’m not sure where would be the best place to run Docker container for public access without cost. I wonder if GitHub pages allows that..

**Comment by manoaman** [comment]
Hi again, 

For the 3d near-isotropic chunks, what exactly do you mean by efficient?

> For viewing multiple orthogonal planes, or for volume rendering, using 3d near-isotropic chunks will be much more efficient.

So if I understand correctly, CloudVolume does not seem to support this.  Is there a software to support this file conversion?

https://github.com/seung-lab/cloud-volume/issues/319#issuecomment-613646448

> By 3d isotropic chunks I just mean a chunk size where all dimensions are approximately the same length in physical units, e.g. if your voxel size is 200nm x 200nm x 1000nm , an example of an isotropic chunk size would be 100 x 100 x 20 voxels, because then the chunk is 20um x 20um x 20um.

**Comment by jbms** [comment]
If you upload the data to google cloud storage, it will be accessible from Neuroglancer directly as long as you enable CORS on the bucket as I described.  You don't need a docker container.  Any other public web server would also work.

You also don't necessarily need to upload the entire dataset --- if the problem can be reproduced just on the coarser resolutions, for example, you could upload just those.

As far as your question about getting "isotropic" chunks --- it is not necessary that they be exactly isotropic, e.g. 1000x1000x1000nm --- it is okay to just be close to isotropic, e.g. 800x800x1000nm.

When I say near-isotropic chunks are more efficient, I mean that the amount of data that must be downloaded in order to display a given view in neuroglancer is lower.

If you use e.g. 512x512x1 chunks, then viewing xy cross sections is very efficient, essentially no overhead.  But viewing an xz or yz cross section will be very inefficient --- 512 times as much data will be required.
